NIFI-8260 Implement an Upload File capability in the NiFi UI for flow definitions.

mtien-apache opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_This feature allows the user to upload a previously exported JSON flow definition into the NiFi UI and it will create a new Process Group, bypassing NiFi Registry.

NIFI-8260 - [WIP] Created Upload File capability for front end and server side logic.
Created a new API endpoint to upload a JSON flow definition and transform it to a Versioned Flow Snapshot.
Fixed the client side file form to reset after submit.
Fixed the canvas to instantly update and show the process group after submitting the file.
Changed the Add Processor Group dialog UI based on design notes.
Changed the Upload File link to an icon and moved to the name input field.
Changed the 'Registry Import' link to say 'Import from Registry' and moved to the bottom of the dialog.
Display the filename when a file is selected.
Added a cancel file button to the Process Group dialog._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

mtien-apache commented:
The PR is ready for initial review. The overall functionality works. I have one more UI enhancement that I'm working on now - it's for disabling the "Add Process Group" dialog "Add" button until the dialog is ready for submission.
cc: @pvillard31 @mcgilman

joewitt commented:
When I read this PR title I was thinking this was a very different thing.  Reading the JIRA I think it is more specific to being able to upload a 'flow definition' which will create that flow on the NiFi it is uploaded to.  Can we please clarify the description?

The term 'flow file' is a specific thing in NiFi and it means 'data in the flow' as represented by a set of attributes and content.

mtien-apache commented:
@joewitt Yes! Changed the term to flow definition.

pvillard31 commented:
Built the PR and gave it a try. When trying to import a flow (json file) that I just exported from somewhere else on the canvas by right clicking on a process group / download flow definition, I always got the below stack trace:

````
2021-02-25 23:42:36,617 ERROR [NiFi Web Server-234] o.a.nifi.web.api.config.ThrowableMapper An unexpected error has occurred: java.lang.NullPointerException. Returning Internal Server Error response.
java.lang.NullPointerException: null
        at org.apache.nifi.web.api.ProcessGroupResource.lambda$uploadProcessGroup$106(ProcessGroupResource.java:4373)
        at org.apache.nifi.web.api.ApplicationResource.withWriteLock(ApplicationResource.java:719)
        at org.apache.nifi.web.api.ProcessGroupResource.uploadProcessGroup(ProcessGroupResource.java:4328)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:148)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:191)
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:200)
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:103)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:493)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:415)
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:104)
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416)
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342)
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229)
        at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1452)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1626)
        at org.apache.nifi.web.filter.RequestLogger.doFilter(RequestLogger.java:66)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
...
````

mtien-apache commented:
@pvillard31 Good catch - I just pushed a commit that should fix this. If uploading a flow that was downloaded from the canvas, the flow does not contain snapshot metadata. I was trying to access the metadata, but it's not necessary. Thanks.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:
```diff
@@ -3766,7 +3765,7 @@ public void updateFlow(final VersionedFlowSnapshot proposedSnapshot, final Strin
         writeLock.lock();
         try {
-            verifyCanUpdate(proposedSnapshot, true, verifyNotDirty);
+             verifyCanUpdate(proposedSnapshot, true, verifyNotDirty);
```
It looks like the changes in this file are just formatting related, should the changes be reverted?

mtien-apache replied:
Strange - I don't think I touched this file. I'll revert it.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java:
```diff
@@ -15,9 +15,35 @@
-
 import com.google.common.collect.Sets;
 import io.prometheus.client.CollectorRegistry;
+import java.io.IOException;
```
Recommend reverting the changes to this file since it is only reordering imports.

mtien-apache replied:
Reverted.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+        }
+
+        if (StringUtils.isBlank(groupId)) {
+            throw new IllegalArgumentException("The parent process group id must be the same as specified in the URI.");
```
The check evaluates whether `groupId` is blank, but the error message seems to indicate that some other check is happening, should the error message be changed, or is some other check necessary?

mtien-apache replied:
Revised the message.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+
+        // get the contents of the InputStream as a String
+        String stringContent = null;
+        if (in != null) {
```
Should a `null` InputStream throw an Exception?  That would allow passing the InputStream directly to `ObjectMapper.readValue()` as opposed to reading it into a String.

mtien-apache replied:
Fixed - a `NullPointerException` can now be thrown.

exceptionfactory replied:
Thanks for adding the `NullPointException`.  To the second point of the comment, it should be possible to remove the call to `IOUtils.toString()` and pass the `InputStream` directly to `ObjectMapper.readValue()`.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+                mapper.setDefaultPropertyInclusion(JsonInclude.Value.construct(JsonInclude.Include.NON_NULL, JsonInclude.Include.NON_NULL));
+                mapper.setAnnotationIntrospector(new JaxbAnnotationIntrospector(mapper.getTypeFactory()));
+                mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
The `ObjectMapper` instance is thread-safe, so it could be configured as a member variable of `ProcessGroupResource` and these configuration methods can be run once instead of on each upload.

mtien-apache replied:
Moved `ObjectMapper` and its configurations to be a member variable.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+
+                deserializedSnapshot = mapper.readValue(stringContent, VersionedFlowSnapshot.class);
+            } catch (JsonParseException jpe) {
+                logger.warn("The file uploaded is not a valid JSON format: ", jpe);
```
The log message seems to indicate that there should be more information, recommend adjusting the invocation:
```suggestion
                logger.warn("Parsing uploaded JSON failed", jpe);
```

mtien-apache replied:
Revised the message.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                deserializedSnapshot = mapper.readValue(stringContent, VersionedFlowSnapshot.class);
+            } catch (JsonParseException jpe) {
+                logger.warn("The file uploaded is not a valid JSON format: ", jpe);
+                String responseJson = "The specified file is not a valid JSON format.";
```
This could be changed to a `private static final String` since it is always the same.

mtien-apache replied:
Revised.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                String responseJson = "The specified file is not a valid JSON format.";
+                return Response.status(Response.Status.OK).entity(responseJson).type("application/json").build();
+            } catch (IOException e) {
+                logger.warn("An error occurred while deserializing the flow version.", e);
```
This error message references a "flow version" but it should be more general.
```suggestion
                logger.warn("Deserialization of uploaded JSON failed", e);
```

mtien-apache replied:
Revised.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                logger.warn("The file uploaded is not a valid JSON format: ", jpe);
+                String responseJson = "The specified file is not a valid JSON format.";
+                return Response.status(Response.Status.OK).entity(responseJson).type("application/json").build();
+            } catch (IOException e) {
```
This `IOException` is logged, but is not wrapped an thrown again.  As a result, following method calls could fail, so this Exception should be wrapped and a new Exception should be thrown.

mtien-apache replied:
Fixed.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4179,227 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+            throw new IOException("The uploaded file was empty.");
+        }
+
+        sanitizeRegistryInfo(deserializedSnapshot.getFlowContents());
+
+        // resolve Bundle info
+        serviceFacade.discoverCompatibleBundles(deserializedSnapshot.getFlowContents());
+
+        // if there are any Controller Services referenced that are inherited from the parent group, resolve those to point to the appropriate Controller Service, if we are able to.
+        serviceFacade.resolveInheritedControllerServices(deserializedSnapshot, groupId, NiFiUserUtils.getNiFiUser());
+
+        // create a new ProcessGroupDTO
+        ProcessGroupDTO processGroupDTO = new ProcessGroupDTO();
+
+        processGroupDTO.setParentGroupId(groupId);
+        processGroupDTO.setName(groupName);
+
+        newProcessGroupEntity.setComponent(processGroupDTO);
+        newProcessGroupEntity.setVersionedFlowSnapshot(deserializedSnapshot);
+
+        // create a PositionDTO
+        PositionDTO positionDTO = new PositionDTO();
+        positionDTO.setX(positionX);
+        positionDTO.setY(positionY);
+        newProcessGroupEntity.getComponent().setPosition(positionDTO);
```
Is it possible to refactor the creation of the ProcessGroupEntity into a separate method?  That would encapsulate the initial verification and building of the object prior to it being processed.

mtien-apache replied:
Good idea - refactored.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/main.css:
```diff
@@ -286,45 +290,125 @@ span.details-title {
+    background-color: transparent;
+}
+
+/*#select-file-button button:hover {*/
```
Should this commented style be retained, or can it be removed?

mtien-apache replied:
Removed.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/header/components/nf-ng-group-component.js:
```diff
@@ -129,17 +133,78 @@
+                        var filenameExtension;
+                        if (!nfCommon.isBlank(filename)) {
+                            filenameExtension = filename.replace(/^.*[\\\/]/, '');
+                            filename = filename.replace(/^.*[\\\/]/, '').replace(/\..*/, '');
```
Should these regular expression replacements be moved into a separate function that describes what is happening?  Otherwise the purpose is not quite clear.

mtien-apache replied:
Revised - I moved to two separate methods describing its functions.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/header/components/nf-ng-group-component.js:
```diff
@@ -304,7 +461,7 @@
                 }).promise();
             }
-        }
+        };
```
Is there a reason for adding this semicolon?

mtien-apache replied:
It's not necessary to have the semicolon, but I thought it would be more complete.

exceptionfactory reviewed (commented):
Thanks for the work on this @mtien-apache!  The overall approach looks good, see individual comments for a few improvements.

mtien-apache commented:
@exceptionfactory Thanks for reviewing! I've addressed your feedback and committed the changes.

mtien-apache commented:
My last commit includes the disabled buttons, which completes this PR and is ready for merge once it's approved. Thanks.
@pvillard31

exceptionfactory commented:
> @exceptionfactory Thanks for reviewing! I've addressed your feedback and committed the changes.

Thanks for making the changes @mtien-apache.  I will take another look following the updates. I noticed that `StandardProcessGroup`, `NiFiServiceFacade`, and `StandardNiFiServiceFacade` still show up with reordered imports, can you remove those changes?

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4187,205 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                        serviceFacade.verifyComponentTypes(versionedFlowSnapshot.getFlowContents());
+                    }
+                },
+                processGroupEntity -> {
+                    final ProcessGroupDTO processGroup = processGroupEntity.getComponent();
+
+                    // set the processor id as appropriate
+                    processGroup.setId(generateUuid());
+
+                    // get the versioned flow
+                    final VersionedFlowSnapshot flowSnapshot = processGroupEntity.getVersionedFlowSnapshot();
+
+                    // create the process group contents
+                    final Revision revision = new Revision((long) 0, clientId, processGroup.getId());
+
+                    ProcessGroupEntity entity = serviceFacade.createProcessGroup(revision, groupId, processGroup);
+
+                    if (flowSnapshot != null) {
+                        final RevisionDTO revisionDto = entity.getRevision();
+                        final String newGroupId = entity.getComponent().getId();
+                        final Revision newGroupRevision = new Revision(revisionDto.getVersion(), revisionDto.getClientId(), newGroupId);
+
+                        // We don't want the Process Group's position to be updated because we want to keep the position where the user
+                        // placed the Process Group. We do not want to use the name of the Process Group that is in the Flow Contents.
+                        // To accomplish this, we call updateProcessGroupContents() passing 'false' for the updateSettings flag, set
+                        // the Process Group name, and null out the position.
+                        flowSnapshot.getFlowContents().setPosition(null);
+                        flowSnapshot.getFlowContents().setName(groupName);
+
+                        entity = serviceFacade.updateProcessGroupContents(newGroupRevision, newGroupId, null, flowSnapshot,
+                                getIdGenerationSeed().orElse(null), false, false, true);
+                    }
+
+                    populateRemainingProcessGroupEntityContent(entity);
+
+                    // generate a 201 created response
+                    String uri = entity.getUri();
+                    return generateCreatedResponse(URI.create(uri), entity).build();
+                }
+        );
```
This block of code looks very similar to the implementation in the `createProcessGroup` method.  Did you consider refactoring both methods to minimize duplication?

mtien-apache replied:
I believe I can partially refactor this block. Will make the change.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4187,205 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+                deserializedSnapshot = MAPPER.readValue(stringContent, VersionedFlowSnapshot.class);
+            } catch (JsonParseException jpe) {
+                logger.warn("Parsing uploaded JSON failed", jpe);
+                return Response.status(Response.Status.OK).entity(INVALID_JSON_RESPONSE).type("application/json").build();
```
The call to `type()` indicates that the string provided to `entity()` should be formatted as JSON, which corresponds to the `@Produces(MediaType.APPLICATION_JSON)` annotation on this method.

On closer review, it looks like the status should be `Response.Status.BAD_REQUEST` to return an HTTP 400 instead of `Response.Status.OK`, otherwise wouldn't the JavaScript handling call success in this case?

Stepping back for a moment, is there a reason for returning a response at this point instead of throwing an exception and having the framework return an appropriately formatted response?

exceptionfactory reviewed (changes requested):
Thanks for making the updates @mtien-apache!  On closer review, there appears to be a good bit of duplication from the `createProcessGroup` method.  I also noted a question about the response status and type in the case of failing to parse the JSON.

pvillard31 commented:
Note sure if this has been addressed but I noticed that: if selecting a flow on the canvas ans clicking on the "Group" icon (to make a process group from the selection), then the icon to upload a JSON file appears when it should just ask for the new process group to create from the selection.

<img width="1407" alt="Screenshot 2021-02-26 at 23 36 54" src="https://user-images.githubusercontent.com/11541012/109346830-ad681980-788b-11eb-9793-891cb90fe332.png">

<img width="450" alt="Screenshot 2021-02-26 at 23 37 38" src="https://user-images.githubusercontent.com/11541012/109346855-b48f2780-788b-11eb-9b06-e8162451f9b7.png">

I don't think this is expected. And if I do select a JSON file at this step (instead of just giving a name for my selection to be added in a process group), the process group for my JSON file will be added instead of putting my selection into a process group.

So... I guess it's just about removing the option to upload the file when we click on the "Group" button. It should only be available when drag-and-dropping a process group on the canvas. Thoughts?

mtien-apache commented:
@pvillard31 Agreed the upload file button should be removed when a user clicks on the "Group" button. Will make the change.

mtien-apache commented:
Just pushed a commit - 
- Hide the Upload File button when grouping components @pvillard31 
- Refactored `uploadProcessGroup` and exception handling @exceptionfactory 
- Refactored `nf-ng-group-component.js` @sardell

Thanks

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4157,163 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+
+        // replicate the request or call serviceFacade.updateProcessGroup
+        if (isReplicateRequest()) {
+            return replicate(HttpMethod.POST, newProcessGroupEntity);
```
Unfortunately, this doesn't work when clustered. Here, it is attempting to replicate a POST method by provided an entity. But the receiving method (this method) does not accept an entity - it expects form data and an input stream. Please take a look at the uploadTemplate() method to see how this is handled there. Rather than calling the `replicate` method, it gets the `requestReplicator` and replicates the request itself.

mtien-apache replied:
@markap14 Thanks for reviewing and catching this. I just pushed a commit that added a second endpoint to handle a cluster replicate request. Can you review again? Thanks!

markap14 reviewed (changes requested):
Thanks for submitting the PR @mtien-apache! It's definitely a great feature that we've needed for a while. The UI looks great. Code looks really good also, but we definitely need to update it to work properly in clustered mode. Otherwise I'm a +1 when that's working!

sardell reviewed (approved):
Nice work, @mtien-apache! I'm also a +1 (non-binding) pending changes to make this work properly in clustered mode.

mtien-apache commented:
@sardell Thanks for reviewing!

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java:
```diff
@@ -4173,6 +4158,264 @@ private void sanitizeRegistryInfo(final VersionedProcessGroup versionedProcessGr
+            }
+    )
+    public Response importProcessGroup(
+            @Context final HttpServletRequest httpServletRequest,
```
Is this parameter necessary? It does not appear to be used.

mtien-apache replied:
Removed - not necessary.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/entity/ProcessGroupUploadEntity.java:
```diff
@@ -0,0 +1,111 @@
+        return id;
+    }
+
+    public void setId(String id) {
```
If this is the ID of the process group, and not the ID of the Upload Entity, I would call this getGroupId/setGroupId.

mtien-apache replied:
Changed to getGroupId and setGroupId.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/entity/ProcessGroupUploadEntity.java:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * @return The client ID
+     */
+    public String getClientId() {
```
The Client ID is never really used on its own like this throughout the data model. Client ID is always part of a RevisionDTO. Recommend changing this to be a RevisionDTO. You can see in the ProcessGroupResource.createProcessGroup (or most any other createXYZ method in the *Resource classes) the pattern that is followed:
```
        if (requestProcessGroupEntity.getRevision() == null || (requestProcessGroupEntity.getRevision().getVersion() == null || requestProcessGroupEntity.getRevision().getVersion() != 0)) {
            throw new IllegalArgumentException("A revision of 0 must be specified when creating a new Process group.");
        }
```

So it enforces that the revision be specified, that it have a version, and that the version be 0, in order to create the object. I'm personally not a huge fan of requiring that the field exists and be set to 0, but it's the pattern that is followed everywhere else, so probably best to be consistent.

mtien-apache replied:
Changed to a RevisionDTO and kept the version to be 0 to follow conventions.

markap14 reviewed (changes requested):
Hey @mtien-apache thanks for the update! I tested it out and it worked great on a clustered environment this time! The only concern that I have now is that some of the conventions that we generally follow with the DTO's/Entities are not really followed. Made specific comments inline. I think it's important to be consistent with the established conventions. Otherwise, all looks good!

mtien-apache commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-client-dto/src/main/java/org/apache/nifi/web/api/entity/ProcessGroupUploadEntity.java:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * @return The process group X-position
+     */
+    public Double getPositionX() {
```
Created a PositionDTO with the x/y positions.

mtien-apache commented:
@markap14 I made revisions from your feedback. Thanks for looking at this in detail!

pvillard31 reviewed (approved):
Went through all of this again, I believe comments from all reviewers have been addressed. This looks good to me and is a super addition that will be incredibly useful! Thanks @mtien-apache, I'm going to merge this to main. Thanks to all reviewers as well!

pvillard31 commented:
Merged (and took care of reverting import ordering changes)
